### PR TITLE
fix amend issues

### PR DESCRIPTION
**Pull request by ouwenkg**
### Changes

* fix amend test
* refactor amend tool

Finally:

test tests::amend_data_test::test_amend_tool ... ok

选区_033

### CODE REVIEW

**Comment by leeyr338** [comment]
Check clippy!

File: cita-executor/core/src/cita_executive.rs [hunk 1]
```diff
@@ -104,14 +95,19 @@ impl<'a, B: DB + 'static> CitaExecutive<'a, B> {
             return Err(ExecutionError::VM(VMError::NotEnoughBaseGas));
         }
 
-        if t.action == Action::AbiStore && !self.transact_set_abi(&t.data) {
-            return Err(ExecutionError::TransactionMalformed(
-                "Account doesn't exist".to_owned(),
-            ));
+        let mut amend = Amend::new(self.state_provider.clone());
+        if t.action == Action::AbiStore {
```
Comment (leeyr338) [inline cita-executor/core/src/cita_executive.rs]: while `AbiStore` is an amend function?

File: cita-executor/core/src/libexecutor/fsm.rs [hunk 2]
```diff
@@ -139,311 +139,305 @@ impl FSM for Executor {
     }
 }
 
-#[cfg(test)]
```
Comment (leeyr338) [inline cita-executor/core/src/libexecutor/fsm.rs]: why delete this test case? I thought I have fix all this.

**Comment by leeyr338** [comment]
> take amend logic out from executive.

This is another task! Would you please let it as an independent  PR?

File: cita-executor/core/src/libexecutor/fsm.rs (hunk 2 above)
Comment (ouwenkg) [reply cita-executor/core/src/libexecutor/fsm.rs]: Noop
```sh
wpf@wpf:~/working/github/ouwenkg/cita/cita-executor$ cargo test -p core-executor
   Compiling core-executor v0.1.0 (/home/wpf/working/github/ouwenkg/cita/cita-executor/core)
error[E0599]: no method named `commit` found for type `std::sync::Arc<std::cell::RefCell<cita_vm::state::state::State<trie_db::TrieDB<cita_db::RocksDB>>>>` in the current scope
   --> cita-executor/core/src/libexecutor/fsm.rs:187:38
    |
187 |                 executed_block.state.commit().expect("commit state");
    |                                      ^^^^^^

error[E0599]: no method named `commit` found for type `std::sync::Arc<std::cell::RefCell<cita_vm::state::state::State<trie_db::TrieDB<cita_db::RocksDB>>>>` in the current scope
   --> cita-executor/core/src/libexecutor/fsm.rs:191:38
    |
191 |                 executed_block.state.commit().expect("commit state");
    |                                      ^^^^^^

error[E0599]: no method named `commit` found for type `std::sync::Arc<std::cell::RefCell<cita_vm::state::state::State<trie_db::TrieDB<cita_db::RocksDB>>>>` in the current scope
   --> cita-executor/core/src/libexecutor/fsm.rs:195:38
    |
195 |                 executed_block.state.commit().expect("commit state");
    |                                      ^^^^^^

error[E0061]: this function takes 2 parameters but 3 parameters were supplied
   --> cita-executor/core/src/libexecutor/fsm.rs:345:24
    |
345 |         executed_block.apply_transaction(
    |                        ^^^^^^^^^^^^^^^^^ expected 2 parameters
```

File: cita-executor/core/src/cita_executive.rs (hunk 1 above)
Comment (ouwenkg) [reply cita-executor/core/src/cita_executive.rs]: It sounds so weird. You are right, separated from amend functionality is better.

**Comment by ouwenkg** [comment]
> > take amend logic out from executive.
> 
> This is another task! Would you please let it as an independent PR?

OK

File: cita-executor/core/src/libexecutor/fsm.rs (hunk 2 above)
Comment (leeyr338) [reply cita-executor/core/src/libexecutor/fsm.rs]: It seems that you hasn't rebase the right code!!
see:
https://github.com/cryptape/cita/blob/remove-gpl/cita-executor/core/src/libexecutor/fsm.rs#L187

File: cita-executor/src/backlogs.rs [hunk 3]
```diff
@@ -411,261 +411,261 @@ pub fn wrap_height(height: usize) -> u64 {
...
-    fn test_wrap_height() {
-        assert_eq!(0, wrap_height(::std::usize::MAX));
-        assert_eq!(
-            ::std::usize::MAX as u64 - 1,
-            wrap_height(::std::usize::MAX - 1)
-        );
-        assert_eq!(2, wrap_height(2));
-    }
-}
+// #[cfg(test)]
```
Comment (kaikai1024) [inline cita-executor/src/backlogs.rs]: Do you run `cargo test` or `cargo test -p core?`

Maybe we should bring back the unit test into CI first, then we fix the unit test step by step. @leeyr338

File: cita-executor/core/src/cita_executive.rs (hunk 1 above)
Comment (kaikai1024) [reply cita-executor/core/src/cita_executive.rs]: Please check it again.
`Set abi` is different of `amend abi`

File: cita-executor/core/src/libexecutor/amend.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
+use std::cell::RefCell;
+use std::fmt;
+use std::sync::Arc;
+
+use cita_types::{H160, H256, U256};
+use cita_vm::state::{State, StateObjectInfo};
+use util::Bytes;
+
+///amend the abi data
```
Comment (kaikai1024) [inline cita-executor/core/src/libexecutor/amend.rs]: Fix the comments togeter.

File: cita-executor/core/src/libexecutor/amend.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    pub fn call_amend_data(
+        &mut self,
+        value: U256,
+        data: Option<Bytes>,
+    ) -> Result<AmendResult, AmendError> {
+        let amend_type = value.low_u32();
+        match amend_type {
+            AMEND_ABI => self.transact_set_abi(&(data.to_owned().unwrap())),
```
Comment (kaikai1024) [inline cita-executor/core/src/libexecutor/amend.rs]: maybe it's better not to use `transact_set/get`. It's nothing about `amend`. Why not use the same function with `amend_type`?

File: cita-executor/src/backlogs.rs (hunk 3 above)
Comment (leeyr338) [reply cita-executor/src/backlogs.rs]: That's a good idea! Can you do this task? I am not familiar with CI.
BTW, I used the command
```shell
cargo test -p core-executor
```
to run Unit test for now.

File: cita-executor/core/src/cita_executive.rs (hunk 1 above)
Comment (ouwenkg) [reply cita-executor/core/src/cita_executive.rs]: Just ensure previous logic, more refactor works later.

File: cita-executor/src/backlogs.rs (hunk 3 above)
Comment (kaikai1024) [reply cita-executor/src/backlogs.rs]: ok

**Pull request closed without merging**